Don't create extra namespaces during pod resize tests

kevinrizza opened the pull request:
#### What type of PR is this?

/kind bug
/kind flake

#### What this PR does / why we need it:

- Modifies the pod resize tests not to create extra unused namespaces
- This reduces load on the namespace controller to help prevent timeouts when other tests are run in parallel
- This is done by pulling the test framework initialization context up so that it is only done before iterating over the test specs rather than creating a separate test framework instance for each test spec. The way the tests were previously structured caused the framework's BeforeEach function to be registered before each test, causing a significant number of extra unused namespaces to be created during these tests.

#### Which issue(s) this PR fixes:
Fixes [#131517](https://github.com/kubernetes/kubernetes/issues/131517)

#### Does this PR introduce a user-facing change?
NONE

kevinrizza commented:
/release-note-none

cici37 commented:
/ok-to-test
/lgtm
/cc @BenTheElder for review when you have time :)

Thank you Kevin for spending time on solving this and appreciate the PR!

cici37 commented:
Note: This will block https://github.com/kubernetes/kubernetes/pull/131514/

BenTheElder reviewed (commented):
/lgtm
/approve

BenTheElder commented:
/hold
for comments from @esotsal

carlory commented:
The framework creates a NS for each IT. This change does not seem to change this fact. It only changes the base name used to generate the NS.

pohly commented:
Each `ginkgo.It` is entirely self-contained. Whatever gets set up for it is torn down after the test and not shared with other tests. As @carlory said, this PR doesn't make any difference in practice. It shares a framework instance, but those are cheap.

Why do you think that creating and tearing down namespaces for each test is a problem? We do that pretty much for all of our tests.

If you absolutely want to do it, then you can define some common setup code in an [ordered container](https://onsi.github.io/ginkgo/#ordered-containers) - but you probably shouldn't.

kevinrizza commented:
@benluddy You're right. And even reducing this test from creating `numberOfTests^2` to `numberOfTests` namespaces still significantly improves the latency of the namespace controller wrt the flake I observed. It would probably be more better if we used a single namespace for all of these tests, but the improvement from that would be negligible by comparison.

kevinrizza committed:
Don't reinitialize test framework during each test case

This commit updates pod resize tests to resolve an
issue where extraneous unused namespaces are created
when the tests are run. This happens because a new
instance of the test framework is generated when
looping over the test specs, registering BeforeEach each
time.

pohly reviewed (approved):
/lgtm
/approve

Now the explanation for this PR makes sense.

esotsal commented:
/hold cancel

BenTheElder commented:
Yeah, we observed this causing flakiness in the ordered namespace deletion tests due to the load, creating one per overall *test* makes sense but this created far more before.

I had spoke to Cici about the context so I focused on reviewing the code diff over non-release-note original description. You can see by inspection how this fixes the issue.

/hold cancel
/triage accepted
/priority important-soon

cici37 commented:
Thank you Kevin for the fix and thank you everyone for the review! ♥️
